### ISSUE TITLE
hello,i have some issue to access on loader.class it maybe cause some particular path or filename

### ISSUE THREAD

**Issue by suguna-mother** [bytedeco/javacpp#323]
java.lang.UnsatisfiedLinkError: no jniGen in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:702)
	at org.bytedeco.javacpp.Loader.load(Loader.java:500)
	at org.bytedeco.javacpp.Loader.load(Loader.java:417)
	at org.bytedeco.javacpp.GenICam3.<clinit>(Gen.java:25)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:472)
	at org.bytedeco.javacpp.Loader.load(Loader.java:417)
	at org.bytedeco.javacpp.Pylon5.<clinit>(Py.java:41)

**Comment by saudet** [comment]
Please try to follow this guide: https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes

**Comment by suguna-mother** [comment]
hai, i have some major issues to connect with libraries of my header file
so can u explain with some model example program ,to set path for library
and i have seen your link: https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes
but i didn't get idea

**Comment by saudet** [comment]
There is a "Hello, World" example here:
https://github.com/bytedeco/helloworld

**Comment by suguna-mother** [comment]
so,how do you connect .dll files with java
can u tell me

**Comment by saudet** [comment]
The main issue right now preventing further automation is the parser, see issue #51.

**Comment by suguna-mother** [comment]
thanks for comment,but mainly i have issue in JNI which i need to set native library path
to access for Loader.load() to get perfect library path(to access class file)
so do u have any idea about to set path in JNI(Native library) and that can be run in netbeans
if u have,please kindly mention it
it almost one exception that i need to solve for my project.
and this is new for me in JavaCPP and it mite need to convert the source c/c++ source to java source code
so in this format i had an exception in native library path to load header file of(c++)

**Comment by saudet** [comment]
What is the error message that you get?

**Comment by suguna-mother** [comment]
an major issue, Is path setting for native library that need to connect to header and .dll files
if i use Loader.load() its an exception of (no FILENAME in library path)
And if i use System.load("FILENAME") it show's an exception of(identify an perfect library path)
and more thing did you ever try for JavaCPP to native library.
if you have an idea just make an note.for me..
it,s been useful

**Comment by saudet** [comment]
Are you still talking about the `UnsatisfiedLinkError`? That just means the library hasn't been built, for example, by using these kinds of commands:
```
$ javac -cp javacpp.jar NativeLibraryConfig.java
$ java -jar javacpp.jar NativeLibraryConfig
$ javac -cp javacpp.jar NativeLibrary.java
$ java -jar javacpp.jar NativeLibrary
```

**Comment by suguna-mother** [comment]
Are you still talking about the UnsatisfiedLinkError?-YES
$ javac -cp javacpp.jar NativeLibraryConfig.java
$ java -jar javacpp.jar NativeLibraryConfig
$ javac -cp javacpp.jar NativeLibrary.java
$ java -jar javacpp.jar NativeLibrary
and i thinking i have issues in classpath in windows while compiling its good,but at runtime its shows
like  cannot find classfile even i fixed classpath in environment variable

**Comment by saudet** [comment]
Try to set the "org.bytedeco.javacpp.logger.debug" system property to "true". It will output more information about what it's trying to do, and the errors it encounters.

**Comment by suguna-mother** [comment]
hi,do you try this helloworld program since i have done this helloworld using netbeans i had same exception
Exception in thread "main" java.lang.UnsatisfiedLinkError: no jnihelloworld in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1483)
	at org.bytedeco.javacpp.Loader.load(Loader.java:1186)
	at org.bytedeco.javacpp.Loader.load(Loader.java:1036)
	at org.bytedeco.helloworld.global.helloworld.<clinit>(helloworld.java:12)
	at org.bytedeco.helloworld.presets.HelloWorldTest.main(HelloWorldTest.java:24)
Caused by: java.lang.UnsatisfiedLinkError: no helloworld in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1483)
	at org.bytedeco.javacpp.Loader.load(Loader.java:1137)
	... 3 more
------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------
Total time: 14.592s
Finished at: Tue Aug 06 12:34:33 IST 2019
Final Memory: 9M/118M
------------------------------------------------------------------------
Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (default-cli) on project helloworld: Command execution failed. Process exited with an error: 1 (Exit value: 1) -> [Help 1]

To see the full stack trace of the errors, re-run Maven with the -e switch.
Re-run Maven using the -X switch to enable full debug logging.

For more information about the errors and possible solutions, please read the following articles:
[Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

**Comment by suguna-mother** [comment]
if you done this helloworld program just please explain

**Comment by saudet** [comment]
Works fine here:
```
git clone https://github.com/bytedeco/javacpp-presets
cd javacpp-presets
git checkout helloworld
cd helloworld
mvn clean install
mvn clean install -f platform -Djavacpp.platform.host
cd samples
mvn clean compile exec:java -Djavacpp.platform.host
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.560 s
[INFO] Finished at: 2019-08-06T16:46:23+09:00
[INFO] ------------------------------------------------------------------------
```

**Comment by suguna-mother** [comment]
and what platform u are using

**Comment by saudet** [comment]
Linux, Mac, and Windows

**Comment by suguna-mother** [comment]
did u run the program in windows

**Comment by saudet** [comment]
Actually, no, it doesn't work on Windows because `libhelloworld-0.dll` isn't getting copied. We need to add it to the `@Platform(..., preload = "libhelloworld-0")` annotation. @matteodg Could you fix this?

**Comment by suguna-mother** [comment]
Ok,then how could i possibly run in windows
and u also shown me the link to solve in mapping index for Exception.
it's any possibility to solve in windows

**Comment by saudet** [comment]
@suguna-mother Just update the annotation in `helloworld/src/main/java/org/bytedeco/helloworld/presets/helloworld.java` like this:
```java
             @Platform(
                     value = {
                         "linux-x86",
                         "linux-x86_64",
                         "macosx-x86_64",
                         "windows-x86",
                         "windows-x86_64"
                     },
                     include = "helloworld.h",
                     link = "helloworld",
                     preload = "libhelloworld-0"
             )
```

**Comment by suguna-mother** [comment]
yes,i have declared in this same model

**Comment by suguna-mother** [comment]
**_**Some problems were encountered while building the effective model for **org.bytedeco:helloworld:jar:1.0-1.5.1-SNAPSHOT**_****
'version' contains an expression but should be a constant. @ org.bytedeco:helloworld:${revision}, C:\Users\admin\Documents\NetBeansProjects\HelloWorldTest\javacpp-presets-helloworld\helloworld\pom.xml, line 14, column 14

It is highly recommended to fix these problems because they threaten the stability of your build.

For this reason, future Maven versions might no longer support building such malformed projects.

Inspecting build with total of 1 modules...
Installing Nexus Staging features:
  ... total of 1 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
                                                                        
------------------------------------------------------------------------
Building JavaCPP Presets for helloworld 1.0-1.5.1-SNAPSHOT****

**Comment by saudet** [comment]
That's also something @matteodg should fix, but this is just a warning. It's not an error or an exception.

**Comment by suguna-mother** [comment]
so in what platform that javacpp presents can be run

**Comment by suguna-mother** [comment]
and do you have any idea about the tools that can support in windows

**Comment by saudet** [comment]
This? https://github.com/bytedeco/javacpp-presets/wiki/Build-Environments#windows-x86-and-x86_64

**Comment by suguna-mother** [comment]
so,in what platform that javacpp currently working for

**Comment by suguna-mother** [comment]
can u help me to run helloworld program in ubuntu platform

**Comment by saudet** [comment]
That works just fine as mentioned above: https://github.com/bytedeco/javacpp/issues/323#issuecomment-518550966

**Comment by suguna-mother** [comment]
are you using 'net beans' or 'eclipse' to develop a project for  Linux platform

**Comment by suguna-mother** [comment]
That works just fine as mentioned above: #323 (comment)
for which platfrom you saying

**Comment by saudet** [comment]
Ubuntu, and try on the command line with Maven first.

**Comment by suguna-mother** [comment]
ok thanks

**Comment by suguna-mother** [comment]
did u use docker

**Comment by suguna-mother** [comment]
and you followed by this link: https://github.com/bytedeco/javacpp-presets/wiki/Build-Environments
for linux

**Comment by saudet** [comment]
That's for the existing presets only, but it's a good guide.

**Comment by suguna-mother** [comment]
now i got output of helloworld
READ -------------------------------------------------
Loaded
J boolean = true
J byte = 56
J char = \u0000
J short = 22275
J int = -2147483648
J long = -9223372036854775808
J AsciiString = Hello World!
J UnicodeString = ↑
WRITE ------------------------------------------------
C boolean = true
C boolean = false
C byte = -128
C byte = 127
C char = \u0000
C char = \uFFFF
C short = -32768
C short = 32767
C int -2147483648
C int 2147483647
C long = -9223372036854775808
C long = 9223372036854775807
C AsciiString = Hello byte array!
C AsciiString = Hello ByteBuffer!
C AsciiString = Hello BytePointer!
C UnicodeString = ���������������������������������������������
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:01 min
[INFO] Finished at: 2019-08-12T10:04:35+05:30
[INFO] ------------------------------------------------------------------------
Unloaded
thank for u information

**Comment by suguna-mother** [comment]
****hi**, while compiling helloworld it's executed.
But,when i tried for same command in pylon it's shows  an POM file cann't found**
mvn clean install -f platform -Djavacpp.platform.host
POM file platform specified with the -f/--file command line argument does not exist
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-readable POM /home/system/javacpp-presets/pylon/platform: /home/system/javacpp-presets/pylon/platform (No such file or directory) @
 @
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project  (/home/system/javacpp-presets/pylon/platform) has 1 error
[ERROR]     Non-readable POM /home/system/javacpp-presets/pylon/platform: /home/system/javacpp-presets/pylon/platform (No such file or directory)
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
**so can u tell me how to solve this error**

**Comment by saudet** [comment]
You'll need to add another pom.xml file like this:
https://github.com/bytedeco/javacpp-presets/tree/helloworld/helloworld/platform

**Comment by matteodg** [comment]
> **_**Some problems were encountered while building the effective model for **org.bytedeco:helloworld:jar:1.0-1.5.1-SNAPSHOT**_****
> 'version' contains an expression but should be a constant. @ org.bytedeco:helloworld:${revision}, C:\Users\admin\Documents\NetBeansProjects\HelloWorldTest\javacpp-presets-helloworld\helloworld\pom.xml, line 14, column 14
> 
> It is highly recommended to fix these problems because they threaten the stability of your build.
> 
> For this reason, future Maven versions might no longer support building such malformed projects.
> 
> Inspecting build with total of 1 modules...
> Installing Nexus Staging features:
> ... total of 1 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
> 
> Building JavaCPP Presets for helloworld 1.0-1.5.1-SNAPSHOT****

@saudet,  I've copied it from existing pom.xml files (I think I copied from cpython/pom.xml), let me know if there is anything else to fix.

I'm creating a new PR with the latest version of helloworld (2.0) with an example of a struct.
Sorry for the delay on my replies!

**Comment by suguna-mother** [comment]
**hi,i have created that pom.xml file**
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>

	<parent>
		<groupId>org.bytedeco</groupId>
		<artifactId>javacpp-presets</artifactId>
		<version>1.3.3-SNAPSHOT</version>
		<relativePath>../../</relativePath>
	</parent>

	<groupId>org.bytedeco</groupId>
	<artifactId>pylon-platform</artifactId>
	<version>5.0.5-${project.parent.version}</version>
	<packaging>jar</packaging>
	<name>JavaCPP Platform for Pylon</name>

	<properties>
		
		<javacpp.moduleId>Pylon</javacpp.moduleId>		
		<revision>1.0-${project.parent.version}</revision>
		<generate-sources.skip>true</generate-sources.skip>
	</properties>

	<dependencies>
			<dependency>
			<groupId>${project.groupId}</groupId>
			<artifactId>${javacpp.moduleId}</artifactId>
			<version>${project.version}</version>		
			</dependency>	
			
			<dependency>		
				<groupId>${project.groupId}</groupId>		
				<artifactId>${javacpp.moduleId}</artifactId>		
				<version>${project.version}</version>		
				<classifier>${javacpp.platform.linux-x86}</classifier>		
			</dependency>
			
			<dependency>		
				<groupId>${project.groupId}</groupId>		
				<artifactId>${javacpp.moduleId}</artifactId>		
				<version>${project.version}</version>		
				<classifier>${javacpp.platform.linux-x86_64}</classifier>		
			</dependency>	
			
			<dependency>		
				<groupId>${project.groupId}</groupId>		
				<artifactId>${javacpp.moduleId}</artifactId>		
				<version>${project.version}</version>		
				<classifier>${javacpp.platform.macosx-x86_64}</classifier>		
			</dependency>	
			
			<dependency>		
				<groupId>${project.groupId}</groupId>		
				<artifactId>${javacpp.moduleId}</artifactId>		
				<version>${project.version}</version>		
				<classifier>${javacpp.platform.windows-x86}</classifier>		
			</dependency>	
			
			<dependency>		
				<groupId>${project.groupId}</groupId>		
				<artifactId>${javacpp.moduleId}</artifactId>		
				<version>${project.version}</version>		
				<classifier>${javacpp.platform.windows-x86_64}</classifier>
			</dependency>
			
	</dependencies>
				
	<build>
		<plugins>
			<plugin>
				<artifactId>maven-jar-plugin</artifactId>
					<executions>	
					
						<execution>		
							<id>default-jar</id>		
							<configuration>		
								<archive>		
									<manifestEntries>		
										<Class-Path>${javacpp.moduleId}.jar ${javacpp.moduleId}-linux-x86.jar ${javacpp.moduleId}-linux-x86_64.jar ${javacpp.moduleId}-macosx-x86_64.jar ${javacpp.moduleId}-windows-x86.jar ${javacpp.moduleId}-windows-x86_64.jar</Class-Path>		
									</manifestEntries>		
								</archive>		
							</configuration>		
						</execution>		
									
						<execution>		
							<id>empty-javadoc-jar</id>		
								<goals>		
									<goal>jar</goal>		
								</goals>		
								
								<configuration>		
									<classifier>javadoc</classifier>		
								</configuration>		
						</execution>	
						
						<execution>		
							<id>empty-sources-jar</id>		
								<goals>		
									<goal>jar</goal>		
								</goals>		
								
								<configuration>		
									<classifier>sources</classifier>		
								</configuration>
								
						</execution>
						
					</executions>
				
			</plugin>
					
			
			<plugin>
					
					<groupId>org.moditect</groupId>
					<artifactId>moditect-maven-plugin</artifactId>
					
					<executions>
						<execution>
							<id>add-module-infos</id>
							<phase>none</phase>
						</execution>
						
						<execution>
							<id>add-platform-module-info</id>
							<phase>packages</phase>
							<goals>
							<goal>add-module-info</goal>
							</goals>
					
							<configuration>
								<modules>
									<module>
										<file>${project.build.directory}/${project.artifactId}.jar</file>		
										<moduleInfoSource>		
											module org.bytedeco.${javacpp.moduleId}.platform {		
											requires org.bytedeco.${javacpp.moduleId}.linux.x86;		
											requires org.bytedeco.${javacpp.moduleId}.linux.x86_64;		
											requires org.bytedeco.${javacpp.moduleId}.macosx.x86_64;		
											requires org.bytedeco.${javacpp.moduleId}.windows.x86;		
											requires org.bytedeco.${javacpp.moduleId}.windows.x86_64;		
											}		
										</moduleInfoSource>		
									</module>		
								</modules>		
							</configuration>		
						</execution>		
					</executions>
			</plugin>		
		</plugins>
	</build>		
		
</project>



> > **_**Some problems were encountered while building the effective model for **org.bytedeco:helloworld:jar:1.0-1.5.1-SNAPSHOT**_****
> > 'version' contains an expression but should be a constant. @ org.bytedeco:helloworld:${revision}, C:\Users\admin\Documents\NetBeansProjects\HelloWorldTest\javacpp-presets-helloworld\helloworld\pom.xml, line 14, column 14
> > It is highly recommended to fix these problems because they threaten the stability of your build.
> > For this reason, future Maven versions might no longer support building such malformed projects.
> > Inspecting build with total of 1 modules...
> > Installing Nexus Staging features:
> > ... total of 1 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
> > Building JavaCPP Presets for helloworld 1.0-1.5.1-SNAPSHOT****
> 
> @saudet, I've copied it from existing pom.xml files (I think I copied from cpython/pom.xml), let me know if there is anything else to fix.
> 
> I'm creating a new PR with the latest version of helloworld (2.0) with an example of a struct.
> Sorry for the delay on my replies!

that,s fine,but i created helloworld in ubuntu
and now i trying to fix in pylon,which shows some error for 
**### Failed to execute goal on project pylon-platform: Could not resolve dependencies for project
org.bytedeco:pylon-platform:jar:5.0.5-1.3.3-SNAPSHOT: The following artifacts could not be resolved:
org.bytedeco:Pylon:jar:5.0.5-1.3.3-SNAPSHOT, org.bytedeco:Pylon:jar:linux-x86_64:5.0.5-1.3.3-
SNAPSHOT: Could not find artifact org.bytedeco:Pylon:jar:5.0.5-1.3.3-SNAPSHOT -> [Help 1]**

**Comment by suguna-mother** [comment]
@saudet ,do you ever execute the pylon program

**Comment by saudet** [comment]
@matteodg Thanks for the update!

@suguna-mother No, but @jpsacha may have.

**Comment by jpsacha** [comment]
I was able to create partial wrapper for Pylon in javacpp-presets. I was able to capture images with it. There are some examples here:
https://github.com/bytedeco/javacv-examples/tree/pylon/Pylon-demo

I was not able to complete the wrapper, got stuck on some GeniCAM C++ constructs. While I was able to capture  images with Pylon wrapper, I was not able to adjust camera parameters (the missing part of the wrapper). 

Due to those problems, I stopped using Pylon. It was easier to create Spinnaker wrapper and use PointGrey/FLIR cameras.
